### PR TITLE
Remove Trainer's `track_grad_norm` argument

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Part of Trainer 2.0 clean up and simplification effort.

Before:

```py
trainer = Trainer(track_grad_norm=2)

# Optionally customize logging in LM:
def log_grad_norm(self, grad_norm_dict):
    self.log(...)
```

Now:

```py
# Override LM hook
def on_before_opimizer_step(self, optimizer):
    # inspect (unscaled) gradients here
    self.log_dict(grad_norm(self, norm_type=2))


```

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda @justusschock @awaelchli

### CODE REVIEW

File: docs/source-pytorch/debug/debugging_intermediate.rst [hunk 1]
```diff
@@ -42,16 +42,26 @@ in the training set to turn off shuffle for you.
...
+One major problem that plagues models is exploding gradients.
+Gradient clipping is one technique that can help keep gradients from exploding.
+
+You can keep an eye on the gradient norm by logging it in your LightningModule:
 
 .. testcode::
 
-    # the 2-norm
-    trainer = Trainer(track_grad_norm=2)
+    from lightning.pytorch.utilities import grad_norm
```
Comment (carmocca) [inline docs/source-pytorch/debug/debugging_intermediate.rst:52]: The PyTorch docs CI job fails here because this triggers the lightning import but `fastapi` is not installed.

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
